### PR TITLE
Use Fontconfig to get the font paths instead of relying on hardcoded values

### LINKED ISSUE #136: [linux] Use fontconfig to locate fonts instead of hardcoding the paths
Currently, LCUI hard-codes the font paths, like here: https://github.com/lc-soft/LCUI/blob/402bc031b424923b89feea390a84a2c15a9b1fdb/src/font/fontlibrary.c#L845

These are not guaranteed to be in the same location across different linux systems.
https://github.com/lc-soft/LCUI/blob/402bc031b424923b89feea390a84a2c15a9b1fdb/src/font/fontlibrary.c#L848 for example, is located in `/usr/share/fonts/noto-cjk/NotoSansCJK-Regular.ttc` on my system.

I suggest using [Fontconfig](https://www.freedesktop.org/wiki/Software/fontconfig/), which would allow to match fonts with their names instead of using the absolute paths.

I am preparing a pull request that updates fontlibrary.c to use Fontconfig for you to review.

### PR DESCRIPTION
**Pull request by coderobe**
Fixes #136.

@lc-soft

### CODE REVIEW

**Comment by coderobe** [comment]
Currently, `fontlibrary.c`'s `DeleteFont( LCUI_Font font )` may lead to a double-free when a font family gets loaded multiple times. I'm still investigating this - don't merge yet

**Comment by coderobe** [comment]
The last commit fixes leaked memory valgrind was complaining about. The only thing left is to ignore adding fonts that have already been loaded, because fontconfig will fall back to the default font when it can't find the one searched for.

eg. loading the fonts `Ubuntu` and `Roboto` may both load a different third font like `Dejavu` when neither are installed on the system. This causes `Dejavu` (or some other font, depending on the system configuration) to be loaded twice, which breaks `DeleteFont( LCUI_Font font )` (double-free).

Because of this, the test crashes on my system with this stacktrace:
```gdb
#0  0x00007ffff6020860 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff6021ec9 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff6063437 in __libc_message () from /usr/lib/libc.so.6
#3  0x00007ffff606978b in malloc_printerr () from /usr/lib/libc.so.6
#4  0x00007ffff606b25c in _int_free () from /usr/lib/libc.so.6
#5  0x00007ffff7badb4d in DeleteFont () from /home/coderobe/Desktop/dev/lcui/src/.libs/libLCUI.so.3
#6  0x00007ffff7baf07d in LCUI_FreeFontLibrary ()
   from /home/coderobe/Desktop/dev/lcui/src/.libs/libLCUI.so.3
#7  0x00007ffff7b8b31a in LCUI_Destroy () from /home/coderobe/Desktop/dev/lcui/src/.libs/libLCUI.so.3
#8  0x0000555555556762 in test_css_parser ()
#9  0x0000555555556302 in main ()
```

File: configure.ac [hunk 1]
```diff
@@ -89,8 +89,8 @@ if test "$want_font_engine" = yes; then
 		AC_CHECK_LIB([freetype], [FT_Init_FreeType], [
 			want_font_engine=yes
 			font_engine_name=freetype2
-			LCUI_LIBS="$LCUI_LIBS `pkg-config --libs freetype2`"
-			CFLAGS="$CFLAGS `pkg-config --cflags-only-I freetype2`"
+			LCUI_LIBS="$LCUI_LIBS `pkg-config --libs freetype2` `pkg-config --libs fontconfig`"
+			CFLAGS="$CFLAGS `pkg-config --cflags-only-I freetype2` `pkg-config --cflags-only-I fontconfig`"
```
Comment (lc-soft) [inline configure.ac]: I think you should add a check for fontconfig, like this:

```
want_font_config = yes
AC_ARG_ENABLE(font-config, AC_HELP_STRING([--enable-font-config],
[turn on font-config [[default=yes]]]), [want_font_config=$enableval])
if test "$want_font_engine" = yes; then
        AC_CHECK_LIB([freetype], [FT_Init_FreeType], []
        ......
        AC_CHECK_LIB([fontconfig], [......], [......])
fi
```

The result of the `./configure` script can be like this:

``` text
Build with lcui-builder support .... : yes
Build with libpng support .......... : yes
Build with libjpeg support ......... : yes
Build with font-engine support ..... : freetype2
Build with font-config support ..... : fontconfig
Build with thread support .......... : pthread
Build with video support ........... : x11
```

File: src/font/fontconfig.c [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * 没有，请查看：<http://www.gnu.org/licenses/>.
+ * ****************************************************************************/
+
+#ifndef LCUI_BUILD_IN_WIN32
+
+#include <stdlib.h>
+#include <string.h>
+#include <fontconfig/fontconfig.h>
+
+char* Fontconfig_Path( char* name )
```
Comment (lc-soft) [inline src/font/fontconfig.c]: I think you should check whether fontconfig is disabled.

``` diff
char* Fontconfig_Path( char* name )
{
+ #ifdef USE_FONT_CONFIG
        ...
+ #else
        return NULL;
+ #endif
}
```

Comment (lc-soft) [reply src/font/fontconfig.c]: I suggest that you change the function name to `Fontconfig_GetPath`.

**Comment by coderobe** [comment]
The last few commits implement your requested changes.

File: src/font/fontlibrary.c [hunk 3]
```diff
@@ -193,6 +193,7 @@ LCUI_Font Font( const char *family_name, const char *style_name )
 
 void DeleteFont( LCUI_Font font )
 {
+	printf("DEBUG: font family at %p (%s) addr %p\n", font->family_name, font->family_name, font);
```
Comment (lc-soft) [inline src/font/fontlibrary.c]: Please remove this debug message.

File: src/font/fontlibrary.c [hunk 4]
```diff
@@ -896,6 +894,11 @@ void LCUI_InitFontLibrary( void )
 			break;
 		}
 	}
+#ifndef LCUI_BUILD_IN_WIN32
+	for( i = 0; i < MAX_FONTFILE_NUM; ++i ) {
+		free( (void*)(fonts[i].path) );
```
Comment (lc-soft) [inline src/font/fontlibrary.c]: You do not need to convert pointer type.

``` diff
- free( (void*)(fonts[i].path) );
+ free( fonts[i].path );
```

**Comment by lc-soft** [comment]
I suggest you merge all commits into one.

```
feat(font): add fontconfig ......
```

File: src/font/fontlibrary.c (hunk 4 above)
Comment (coderobe) [reply src/font/fontlibrary.c]: Not converting the pointer type results in a compiler warning.
```diff
fontlibrary.c: In function ‘LCUI_InitFontLibrary’:
fontlibrary.c:898:9: warning: passing argument 1 of ‘free’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
>   free( fonts[i].path );
>         ^~~~~
In file included from fontlibrary.c:42:0:
/usr/include/stdlib.h:448:13: note: expected ‘void *’ but argument is of type ‘const char *’
> extern void free (void *__ptr) __THROW;
>             ^~~~
```

**Comment by coderobe** [comment]
Are you sure you want the commits to be squashed? That way we will lose history about which commit(s) introduced what.

**Comment by coderobe** [comment]
I have squashed most of the commits down into these 4. This should be better than before.

File: src/font/fontlibrary.c [hunk 5]
```diff
@@ -834,31 +834,28 @@ void LCUI_InitFontLibrary( void )
 		{ FONTDIR"msyh.ttc", "Microsoft YaHei", NULL }
 	};
 #else
-#define FONTDIR "/usr/share/fonts/"
-#define MAX_FONTFILE_NUM 4
+#define MAX_FONTFILE_NUM 3
 	struct {
```
Comment (lc-soft) [inline src/font/fontlibrary.c]:
``` diff
+ #elif defined(USE_FONTCONFIG)
+			Fontconfig_GetPath( "Ubuntu" ),
+ #else
			FONTDIR"truetype/ubuntu-font-family/Ubuntu-R.ttf",
+ #endif
```

File: src/font/fontlibrary.c [hunk 6]
```diff
@@ -896,6 +893,11 @@ void LCUI_InitFontLibrary( void )
 			break;
 		}
 	}
+#ifndef LCUI_BUILD_IN_WIN32
```
Comment (lc-soft) [inline src/font/fontlibrary.c]:
``` diff
- #ifndef LCUI_BUILD_IN_WIN32
+ #ifdef USE_FONTCONFIG
```

File: src/font/fontlibrary.c [hunk 7]
```diff
@@ -834,31 +834,28 @@ void LCUI_InitFontLibrary( void )
...
-			FONTDIR"truetype/wqy/wqy-microhei.ttc",
+			Fontconfig_GetPath( "WenQuanYi Micro Hei" ),
 			"WenQuanYi Micro Hei", NULL
 		}
 	};
 #endif
 
 	fontlib.font_cache_num = 1;
 	fontlib.font_cache = NEW( LCUI_FontCache, 1 );
+	ZEROSET( fontlib.font_cache, LCUI_FontCache );
```
Comment (lc-soft) [inline src/font/fontlibrary.c]: Do not need this code, Because `fontlib.font_cache[0] = FontCache();` will fill data.

**Comment by coderobe** [comment]
note: i'm still interested in this and will continue working on this at some point, but i ran into an issue with autotools not properly defining certain flags and very limited time to work on it right now. just letting you know

**Comment by lc-soft** [comment]
I have manually merged your branch's code.

**Pull request closed without merging**